src/engine/routing_algorithms/routing_base.cpp

```diff
@@ -33,6 +33,75 @@ bool needsLoopBackwards(const PhantomNodes &phantoms)
+        if (source_phantom.forward_segment_id.id == path.front())
+        {
+            //       ............       <-- calculateEGBAnnotation returns distance from 0 to 3
+            //       -->s               <-- subtract offset to start at source
+            //          .........       <-- want this distance as result
+            // entry 0---1---2---3---   <-- 3 is exit node
+            distance -= source_phantom.GetForwardDistance();
+        }
+        else if (source_phantom.reverse_segment_id.id == path.front())
+        {
+            //       ............    <-- calculateEGBAnnotation returns distance from 0 to 3
+            //          s<-------    <-- subtract offset to start at source
+            //       ...             <-- want this distance
+            // entry 0---1---2---3   <-- 3 is exit node
+            distance -= source_phantom.GetReverseDistance();
+        }
+        if (target_phantom.forward_segment_id.id == path.back())
+        {
+            //       ............       <-- calculateEGBAnnotation returns distance from 0 to 3
+            //                   ++>t   <-- add offset to get to target
+            //       ................   <-- want this distance as result
+            // entry 0---1---2---3---   <-- 3 is exit node
+            distance += target_phantom.GetForwardDistance();
+        }
+        else if (target_phantom.reverse_segment_id.id == path.back())
+        {
+            //       ............       <-- calculateEGBAnnotation returns distance from 0 to 3
+            //                   <++t   <-- add offset to get from target
+            //       ................   <-- want this distance as result
+            // entry 0---1---2---3---   <-- 3 is exit node
+            distance += target_phantom.GetReverseDistance();
+        }
+    }
+    else
+    {
+        // there is no shortcut to unpack. source and target are on the same EBG Node.
+        // if the offset of the target is greater than the offset of the source, subtract it
+        if (target_phantom.GetForwardDistance() > source_phantom.GetForwardDistance())
+        {
+            //       --------->t        <-- offsets
+            //       ->s                <-- subtract source offset from target offset
+            //         .........        <-- want this distance as result
+            // entry 0---1---2---3---   <-- 3 is exit node
+            distance = target_phantom.GetForwardDistance() - source_phantom.GetForwardDistance();
+        }
+        else
+        {
+            //               s<---      <-- offsets
+            //         t<---------      <-- subtract source offset from target offset
+            //         ......           <-- want this distance as result
+            // entry 0---1---2---3---   <-- 3 is exit node
+            distance = target_phantom.GetReverseDistance() - source_phantom.GetReverseDistance();
+        }
+    }
+
+    // guard against underflow errors caused by rounding
```

Comment:
Should there be an assert here checking for an epsilon making sure it's _only_ due to rounding and not due to bugs in our code? Otherwise a negative erroneous distance will be silently ignored here.